File: ivy/ivy-compiler/src/cvm/optimize.ts


export function optimize(instructions: string[]): string[] {
    let instructionsString = " " + instructions.join(" ")
    let optimizedInstructions = instructionsString
    .replace(/ 0 ROLL/g, "")
    .replace(/ 0 PICK/g, " DUP")
    .replace(/ 1 ROLL/g, " SWAP")
    .replace(/ 1 PICK/g, " OVER")
    .replace(/ 2 ROLL/g, " ROT")
    .replace(/ TRUE VERIFY/g, "")
    .replace(/ TRUE VERIFY/g, "")
    .replace(/ TRUE VERIFY/g, "")
    .replace(/ SWAP SWAP/g, "")
    .replace(/ OVER OVER/g, " 2DUP")
    .replace(/ SWAP OVER/g, " TUCK")
    .replace(/ DROP DROP/g, " 2DROP")
    .replace(/ SWAP DROP/g, " NIP")
    .replace(/ 5 ROLL 5 ROLL/g, " 2ROT")
    .replace(/ 3 PICK 3 PICK/g, " 2OVER")
    .replace(/ 3 ROLL 3 ROLL/g, " 2SWAP")
    .replace(/ 2 PICK 2 PICK 2 PICK/g, " 3DUP")
    .replace(/ 1 ADD/g, " 1ADD")
    .replace(/ 1 (\d+) ADD/g, " 1ADD")
    .replace(/ 1 SUB/g, " 1SUB")
    .replace(/ SWAP (EQUAL|ADD|BOOLAND|BOOLOR|MIN|MAX)/g, " $1")
    .replace(/ DUP 2 PICK (EQUAL|ADD|BOOLAND|BOOLOR|MIN|MAX)/g, " 2DUP $1")
    .replace(/ EQUAL VERIFY/g, " EQUALVERIFY")
    .replace(/ SWAP TXSIGHASH ROT/g, " TXSIGHASH SWAP")

    return optimizedInstructions.slice(1).split(" ")
}

File: exp/ivy/compiler/optimize.go
package compiler

import "strings"

var optimizations = []struct {
	before, after string
}{
	{"0 ROLL", ""},
	{"0 PICK", "DUP"},
	{"1 ROLL", "SWAP"},
	{"1 PICK", "OVER"},
	{"2 ROLL", "ROT"},
	{"TRUE VERIFY", ""},
	{"SWAP SWAP", ""},
	{"OVER OVER", "2DUP"},
	{"SWAP OVER", "TUCK"},
	{"DROP DROP", "2DROP"},
	{"SWAP DROP", "NIP"},
	{"5 ROLL 5 ROLL", "2ROT"},
	{"3 PICK 3 PICK", "2OVER"},
	{"3 ROLL 3 ROLL", "2SWAP"},
	{"2 PICK 2 PICK 2 PICK", "3DUP"},
	{"1 ADD", "1ADD"},
	{"1 SUB", "1SUB"},
	{"EQUAL VERIFY", "EQUALVERIFY"},
	{"SWAP TXSIGHASH ROT", "TXSIGHASH SWAP"},
	{"SWAP EQUAL", "EQUAL"},
	{"SWAP EQUALVERIFY", "EQUALVERIFY"},
	{"SWAP ADD", "ADD"},
	{"SWAP BOOLAND", "BOOLAND"},
	{"SWAP BOOLOR", "BOOLOR"},
	{"SWAP MIN", "MIN"},
	{"SWAP MAX", "MAX"},
	{"DUP 2 PICK EQUAL", "2DUP EQUAL"},
	{"DUP 2 PICK EQUALVERIFY", "2DUP EQUALVERIFY"},
	{"DUP 2 PICK ADD", "2DUP ADD"},
	{"DUP 2 PICK BOOLAND", "2DUP BOOLAND"},
	{"DUP 2 PICK BOOLOR", "2DUP BOOLOR"},
	{"DUP 2 PICK MIN", "2DUP MIN"},
	{"DUP 2 PICK MAX", "2DUP MAX"},
}

func optimize(opcodes string) string {
	opcodes = " " + opcodes + " "
	looping := true
	for looping {
		looping = false
		for _, o := range optimizations {
			before := " " + o.before + " "
			var after string
			if o.after == "" {
				after = " "
			} else {
				after = " " + o.after + " "
			}
			newOpcodes := strings.Replace(opcodes, before, after, -1)
			if newOpcodes != opcodes {
				looping = true
				opcodes = newOpcodes
			}
		}
	}
	return strings.TrimSpace(opcodes)
}
